Metric Observer instrument specification (refinement)

jmacd opened the pull request:
This is a clarification of the format OTEP 0008.  This topic encountered last-minute debate in https://github.com/open-telemetry/opentelemetry-specification/pull/250 and was left out. The section on calling conventions and examples here should clarify the issue.

yurishkuro commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,184 @@
+exporter, provided it does not contradict the requirement above.
+
+Callbacks should avoid calling OpenTelemetry APIs other than the
+interface provided to `Observe().  This prevents the SDK from
```
```suggestion
interface provided to `Observe()`.  This prevents the SDK from
```

jmacd replied:
Fixed, thanks.

yurishkuro commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,184 @@
+        cpuTempByCore[i] = cpuTemp.Bind(meter.createLabelSet("core", i));
+    }
+
+    observer.setCallback(
```
what is `observer` here?

jmacd replied:
Fixed. This should have been `cpuTemp`.

jmacd commented:
@yurishkuro I've added two paragraphs to try to address your question. I will confess to having mixed feelings about this API. I volunteered to write this specification after I removed it from the metrics spec earlier, in an agreement with @bogdandrutu.  My original draft in OTEP 0008 didn't talk much about about calling conventions, but I had not at that time been thinking of supporting "bound" observer instruments. The addition of bound observer instruments raises some questions (e.g., what prevents you from calling them in a _different_ observer's callback?). 

Another possibility (@bogdandrutu please comment) would be to support a callback for unbound observers and individual callbacks for each bound observer.  Why support one callback per instrument, not one callback per bound instrument IOW?

jmacd commented:
(I've asked @bogdandrutu this in the past, and he expressed a preference for a single-callback as proposed here, not a callback per bound observer.  It's fewer allocations.)

bogdandrutu commented:
@jmacd it is not necessary about fewer allocation but also that a system (or external) call like (getting memory) may return data for more than one bound which means that if a callback per-bound is allocated the system call will be done multiple times. In case multiple system (or external) calls are required the callback can deal with that and update multiple bounds. As an example see https://github.com/open-telemetry/opentelemetry-java/blob/master/contrib/runtime_metrics/src/main/java/io/opentelemetry/contrib/metrics/runtime/GarbageCollector.java#L59 where I pre-allocate a list of bounds and update all of them after doing only one system/external call.

bogdandrutu commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,200 @@
+This proposal restores the ability to use callbacks only with a
+dedicated `Observer` kind of instrument with the same semantics as
+Gauge instruments.  Like a Gauge instrument, Observer instruments are
+used to report the current value of a variable.
```
+1

bogdandrutu commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,200 @@
+the "last value" may be interpreted directly as a sum.  Systems with
+support for rate calculations over current sums (e.g., Prometheus)
+will be able to use these metrics directly.  Systems not designed to
+calculate rates from sums may not benefit from this treatment.
```
If really needed we can calculate rates at the client and export them `(current_value - last_value) / collection_interval`.

jmacd replied:
I decided it's best to remove the confusing statement ("Systems not designed...").

bogdandrutu commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,200 @@
+does not execute with any distributed context.  
+
+Whereas Gauge values do have context at the moment `Set()` is called,
+Observer callbacks do not.
```
In other words they do not belong to a specific request. (e.g. process cpu-usage)

jmacd replied:
I added "Observer instruments are appropriate for reporting metrics that are not request specific."

bogdandrutu commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,200 @@
+          @Override
+          public void update(Result result) {
+            for (int i = 0; i < NUM_CORES; i++) {
+              cpuTempByCore[i].Observe(getCPUTemp(i), result);
```
I would design this API a bit different:

`result.Observe(getCPUTemp(i), cpuTempByCore[i])` - if no overloading support in the language then `ObserverWithBound` can be used for the name of the method/function.

jmacd replied:
Done.

jmacd commented:
Please see Bogdan's example (also linked above):
https://github.com/open-telemetry/opentelemetry-java/blob/master/contrib/runtime_metrics/src/main/java/io/opentelemetry/contrib/metrics/runtime/GarbageCollector.java#L59

c24t commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,202 @@
+BoundInstrument)`.  Callbacks MUST use bound instruments corresponding the
+Observer instrument for which they are registered.  It is an error if
+`Result.ObserveBound(value, BoundInstrument)` is called for a Result
+corresponding to a different Observer instrument.
```
As I understand it, "bound" metric instruments are an optimization to avoid repeatedly updating a metric for a particular `labelSet`. This saves us the cost of constructing the `labelSet` and looking up the aggregation to update, since the bound instrument has a reference to the aggregation.

Does this optimization still make sense in the context of callbacks? For non-observer metrics, we wouldn't expect the user to create a bound instrument for a single call.

jmacd replied:
You ask a good question (also below), because if we had one callback per bound Observer/LabelSet, then there's no need to actually expose a Bound Instrument type. There would be a `Observer.Bind(LabelSet, callback)` method to bind a callback to a LabelSet instead, and a callback would be called once per bound label set and it could return a value instead of calling `ObserverResult` callback. This is "case (b)" in the updated language. Potentially this approach can evaluate callbacks in parallel, although that could also cause problems.

This confusion is what led to leaving Observers out of the v0.2 spec. I came away thinking that one callback per bound instrument makes sense, but if we intend a direct calling convention for Observer instruments, then having one callback per instrument gives a smaller surface area (and simpler SDK).  I'm on the fence here.

bogdandrutu replied:
I would suggest one callback per instrument, the main reason is that you can simply implement a callback that updates multiple labelsets, but if you have a callback per labelset and you need to access a system information you may end up doing a lot of unnecessary system calls. See the java example for memory.

I think @c24t is asking to completely remove bounds from the observer metrics because the benefit of bounding is unclear if the callback is called rarely (once per export interval) which I agree with and we can always add the bounds later to the Observers if we prove the performance wrong.

jmacd replied:
It's a good point. The point of the bound instrument is to avoid a map lookup for each operation, but that map lookup isn't needed for observer instruments--they can bypass the map of current records. They observer instrument's observations are output directly, and since the Batcher (in the current arch diagram--I want to rename this) is required to support duplicates I expect no benefit from bound observer instruments.

jmacd replied:
I have removed "bound" observer instruments based on this reasoning.

yurishkuro replied:
I think it would be a move in the wrong direction. IMO, the callback use case is far more obscure than bound instruments (which are used all over the place in Jaeger code, for example). Bound instruments allow the code to provide measurements on its own pace, usually based on the business traffic. If callback is the only way to provide those, then user code needs to be accumulating the measurements between callbacks, a serious complication.

jmacd replied:
The (now removed) case of a bound observer instrument was not meant as a stand-in for the user to accumulate measurements, but for a group of fixed label sets to be repeatedly recorded on the collection interval.

I agree completely with your statement -- Observer instruments should not be used when it involves the user accumulating measurements.  There should not be any benefit or performance advantage for the user because the SDK is designed and heavily optimized to do just this -- accumulate measurements. Bound instruments should be used when recording measurements. 

Observer instruments are meant for "computed values", which are passively collected measurements, not accumulated by the user but computed on the collection interval. The observation (which I credit to @c24t above) is that binding a label set with an observer instrument does not offer any performance benefit, because the emitted values will be directly output to the export pipeline.  They are not buffered or accumulated in the SDK, therefore binding them does no good.

I'm leaving this issue resolved, since we have removed bound observer instruments from this proposal.

c24t commented on text/0072-metric-observer.md:
````diff
@@ -0,0 +1,202 @@
+        });
+  }
+}
+```
````
It's not clear from the text why the user shouldn't be able to set a callback on a bound instrument directly. This seems like a significant difference between `Set` and `Observe`.

jmacd replied:
I tried to add a justification, but I feel about 50/50 on the pros/cons of the one-callback-per-instrument decision.

c24t replied:
Much clearer (even if it's a clear expression of ambiguity about the feature). Thanks!

jmacd commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,202 @@
+
+### Observer calling conventions
+
+Observer callbacks accept a Result interface, which supports both
```
Yes. I've tried to clarify this with more explicit language.

bogdandrutu commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,183 @@
+        new ObserverDouble.Callback<ObserverDouble.Result>() {
+          @Override
+          public void update(Result result) {
+              result.Observe(getCPULoad(), meter.createLabelSet("low_power", isLowPowerMode()));
```
You may want to show that LabelSet can be cached :)

jmacd replied:
Done

jmacd commented:
@bogdandrutu @c24t @yurishkuro please take another look.

yurishkuro commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,184 @@
+convention is not needed for Observer instruments because there is
+little if any performance benefit in doing so (as Observer instruments
+are called during collection, there is no need to maintain "active"
+records concurrent with collection).
```
Bounding an instrument may involve encoding the label set into a wire-ready data structure, which in itself can be expensive if it needs to be done repeatedly on every measurement.

jmacd replied:
The `LabelSet` value is meant to capture any encoding of label sets into wire-ready data structures, since these can be used for more than one bound instrument. In the example snippet, the label set is re-used to emphasize this. The only optimization that I can see would be if the Observer were to output more than once for the same LabelSet per collection period, in which case binding them would support the optimization.

yurishkuro commented on text/0072-metric-observer.md:
```diff
@@ -0,0 +1,184 @@
+  private static final ObserverDouble cpuLoad = ...;
+
+  void init() {
+    LabelSet labelSet = meter.createLabelSet("low_power", isLowPowerMode());
```
compile error, missing `final` :-)
